### PR TITLE
Corfu cluster performance benchmark

### PR DESCRIPTION
**Pull request by xnull**
## Overview

Description:
Measures performance characteristics of a Corfu cluster running in different modes.
Docs: https://github.com/CorfuDB/CorfuDB/blob/corfu_cluster_performance/docs/benchmarks/corfu-cluster.md

Why should this be merged: 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: .mvn/wrapper/maven-wrapper.properties [hunk 1]
```diff
@@ -1 +1 @@
-distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.6.0/apache-maven-3.6.0-bin.zip
\ No newline at end of file
+distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.6.2/apache-maven-3.6.2-bin.zip
```
Comment (vjeko) [inline .mvn/wrapper/maven-wrapper.properties]: Missing newline.

File: benchmarks/src/jmh/java/org/corfudb/benchmarks/cluster/ClusterBenchmark.java [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Stream;
+
+@Slf4j
+public class ClusterBenchmark {
```
Comment (vjeko) [inline benchmarks/src/jmh/java/org/corfudb/benchmarks/cluster/ClusterBenchmark.java]: Can you please explain what the test does? What is its purpose?

File: benchmarks/src/jmh/resources/vm.properties [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+vm2.HostSystem=host-2148
+vm2.Datastore=datastore-2149
+vm2.ResourcePool=resgroup-2147
+vm2.Folder=group-v155
+
+vm3.HostSystem=host-2155
+vm4.Datastore=datastore-2156
+vm4.ResourcePool=resgroup-2154
+vm4.Folder=group-v155
```
Comment (vjeko) [inline benchmarks/src/jmh/resources/vm.properties]: Newline.

File: benchmarks/src/main/java/org/corfudb/benchmarks/util/DataUnit.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
+package org.corfudb.benchmarks.util;
+
+public enum DataUnit {
```
Comment (vjeko) [inline benchmarks/src/main/java/org/corfudb/benchmarks/util/DataUnit.java]: Can't we use something along the lines of storage-units?

File: benchmarks/src/main/resources/logback.xml [hunk 5]
```diff
@@ -13,7 +13,7 @@
 
     <logger name="ch.qos.logback" level="off" additivity="false"/>
 
-    <root level="info">
+    <root level="error">
         <appender-ref ref="stdout"/>
     </root>
 </configuration>
```
Comment (vjeko) [inline benchmarks/src/main/resources/logback.xml]: Newline.

File: checks.xml [hunk 6]
```diff
@@ -235,11 +235,6 @@
         </module>
         <module name="OverloadMethodsDeclarationOrder"/>
         <module name="VariableDeclarationUsageDistance"/>
-        <module name="CustomImportOrder">
```
Comment (vjeko) [inline checks.xml]: Why was this removed?

File: checks.xml [hunk 7]
```diff
@@ -264,25 +259,22 @@
             <property name="allowSamelineMultipleAnnotations" value="true"/>
         </module>
         <module name="NonEmptyAtclauseDescription"/>
-        <module name="JavadocTagContinuationIndentation"/>
+        <module name="JavadocTagContinuationIndentation">
+            <property name="severity" value="ignore"/>
+        </module>
         <module name="SummaryJavadoc">
-            <property name="forbiddenSummaryFragments"
```
Comment (vjeko) [inline checks.xml]: Same here. Why?

File: checks.xml [hunk 8]
```diff
@@ -264,25 +259,22 @@
...
+            <property name="severity" value="ignore"/>
         </module>
-        <module name="JavadocParagraph"/>
         <module name="AtclauseOrder">
             <property name="tagOrder" value="@param, @return, @throws, @deprecated"/>
             <property name="target"
                       value="CLASS_DEF, INTERFACE_DEF, ENUM_DEF, METHOD_DEF, CTOR_DEF, VARIABLE_DEF"/>
         </module>
         <module name="JavadocMethod">
-            <property name="scope" value="public"/>
```
Comment (vjeko) [inline checks.xml]: Can you please revert these changes.

File: it/src/main/java/org/corfudb/universe/UniverseManager.java [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+import org.corfudb.universe.universe.docker.DockerUniverse;
+import org.corfudb.universe.universe.process.ProcessUniverse;
+import org.corfudb.universe.universe.vm.ApplianceManager;
+import org.corfudb.universe.universe.vm.VmUniverse;
+import org.corfudb.universe.universe.vm.VmUniverseParams;
+
+import java.util.function.Consumer;
+
+@Builder
+public class UniverseManager {
```
Comment (vjeko) [inline it/src/main/java/org/corfudb/universe/UniverseManager.java]: You need comments.

File: it/src/main/java/org/corfudb/universe/UniverseManager.java [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+                .testName(testName)
+                .fixture(fixture)
+                .build()
+                .init();
+    }
+
+    @Slf4j
+    @Builder
+    @Getter
+    public static class UniverseWorkflow<T extends Fixture<UniverseParams>> {
```
Comment (vjeko) [inline it/src/main/java/org/corfudb/universe/UniverseManager.java]: Same here. Comments.

File: it/src/main/java/org/corfudb/universe/node/server/CorfuServerParams.java [hunk 11]
```diff
@@ -14,52 +14,68 @@
...
     private final Duration stopTimeout = Duration.ofSeconds(1);
 
+    @Default
+    private final Optional<ContainerResources> containerResources = Optional.empty();
+
+    /**
+     * Corfu server version, for instance: 0.3.0-SNAPSHOT
+     */
+    @NonNull
+    private final String serverVersion;
```
Comment (vjeko) [inline it/src/main/java/org/corfudb/universe/node/server/CorfuServerParams.java]: We already have this parameter in CorfuClusterParams. Can't we just use that one?

File: it/src/main/java/org/corfudb/universe/node/server/CorfuServerParams.java [hunk 12]
```diff
@@ -14,52 +14,68 @@
...
+
+    @NonNull
+    @Default
+    private final Path serverJarDirectory = Paths.get("target");
```
Comment (vjeko) [inline it/src/main/java/org/corfudb/universe/node/server/CorfuServerParams.java]: Can you make this configurable? That would be pretty useful.

File: it/src/main/java/org/corfudb/universe/node/server/SupportServerParams.java [hunk 13]
```diff
@@ -37,11 +35,12 @@
...
     @NonNull
     @Getter
-    private final Level logLevel = Level.DEBUG;
+    private final Level logLevel = Level.INFO;
 
+    @Default
     @NonNull
     @Getter
-    private final NodeType nodeType;
+    private final NodeType nodeType = NodeType.METRICS_SERVER;
```
Comment (vjeko) [inline it/src/main/java/org/corfudb/universe/node/server/SupportServerParams.java]: SupportServer is meant for way more than just metrics. This being a default value makes no sense.

File: it/src/main/java/org/corfudb/universe/node/server/process/CorfuProcessManager.java [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.universe.node.server.CorfuServerParams;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+@Slf4j
+public class CorfuProcessManager {
```
Comment (vjeko) [inline it/src/main/java/org/corfudb/universe/node/server/process/CorfuProcessManager.java]: Comments.

File: it/src/main/java/org/corfudb/universe/node/server/process/ExecutionHelper.java [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+    private ExecutionHelper() {
+        //prevent creating class instances
+    }
+
+    public static ExecutionHelper getInstance() {
+        return INSTANCE;
+    }
+
+    /**
+     * Copy a file
```
Comment (vjeko) [inline it/src/main/java/org/corfudb/universe/node/server/process/ExecutionHelper.java]: More documentation. Is this a local file copy?

File: it/src/main/java/org/corfudb/universe/node/server/process/ExecutionHelper.java [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+        log.info("Copying {} to {}", srcFile, targetFile);
+        copy.execute();
+    }
+
+    /**
+     * Execute a shell command
+     *
+     * @param command shell command
+     */
+    public void executeCommand(Optional<Path> workDir, String command) throws IOException {
```
Comment (vjeko) [inline it/src/main/java/org/corfudb/universe/node/server/process/ExecutionHelper.java]: Ideally, this should be more robust -- how do we know if the execution succeeded?

File: it/src/main/java/org/corfudb/universe/node/server/process/ProcessCorfuServer.java [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+            throw new NodeException("Execution error. Cmd: " + cmdLine, e);
+        }
+    }
+
+    /**
+     * @return the IpAddress of this VM.
+     */
+    @Override
+    public String getIpAddress() {
+        return "127.0.0.1";
```
Comment (vjeko) [inline it/src/main/java/org/corfudb/universe/node/server/process/ProcessCorfuServer.java]: Why are we returning localhost? Also, make this a constant?

File: it/src/main/java/org/corfudb/universe/node/server/process/ProcessCorfuServer.java [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+        try {
+            executeCommand(Optional.empty(), processManager.stopCommand());
+        } catch (Exception e) {
+            String err = String.format("Can't STOP corfu: %s. Process not found", params.getName());
+            throw new NodeException(err, e);
+        }
+    }
+
+    /**
+     * Kill the {@link CorfuServer} process on the VM directly.
```
Comment (vjeko) [inline it/src/main/java/org/corfudb/universe/node/server/process/ProcessCorfuServer.java]: Does not have to be a VM.

File: it/src/main/java/org/corfudb/universe/node/server/process/ProcessCorfuServer.java [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+        try {
+            removeAppDir();
+        } catch (Exception e) {
+            throw new NodeException("Can't clean corfu directories", e);
+        }
+    }
+
+    /**
+     * Remove corfu server application dir.
+     * AppDir is a directory that contains corfu-infrastructure jar file and could have log files, stream-log files and
```
Comment (vjeko) [inline it/src/main/java/org/corfudb/universe/node/server/process/ProcessCorfuServer.java]: Comment formatting. Please wrap your lines.

File: it/src/main/java/org/corfudb/universe/scenario/fixture/FixtureUtil.java [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+import org.corfudb.universe.node.server.vm.VmCorfuServerParams.VmCorfuServerParamsBuilder;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+@Builder
+public class FixtureUtil {
```
Comment (vjeko) [inline it/src/main/java/org/corfudb/universe/scenario/fixture/FixtureUtil.java]: Comments.

File: it/src/main/java/org/corfudb/universe/scenario/fixture/Fixtures.java [hunk 21]
```diff
@@ -57,149 +60,133 @@ private TestFixtureConst() {
...
+            FixtureUtil fixtureUtil = fixtureUtilBuilder.build();
+            ImmutableList<CorfuServerParams> serversParams = fixtureUtil.buildVmServers(
+                    clusterParams, servers, vmPrefix
+            );
+            serversParams.forEach(clusterParams::add);
 
-                serversParams.add(serverParam);
+            ConcurrentMap<String, String> vmIpAddresses = new ConcurrentHashMap<>();
+            for (int i = 0; i < clusterParams.getNumNodes(); i++) {
+                vmIpAddresses.put(vmPrefix + (i + 1), "0.0.0.0");
```
Comment (vjeko) [inline it/src/main/java/org/corfudb/universe/scenario/fixture/Fixtures.java]: Please put this into a ANY_ADDRESS constant.

File: it/src/main/java/org/corfudb/universe/universe/process/ProcessUniverse.java [hunk 22]
```diff
@@ -1,42 +1,70 @@
...
-    private static final UnsupportedOperationException NOT_IMPLEMENTED =
-            new UnsupportedOperationException("Not implemented");
+import java.util.concurrent.atomic.AtomicBoolean;
 
-    @Override
-    public Universe deploy() {
-        throw NOT_IMPLEMENTED;
-    }
+@Slf4j
+public class ProcessUniverse extends AbstractUniverse<NodeParams, UniverseParams> {
```
Comment (vjeko) [inline it/src/main/java/org/corfudb/universe/universe/process/ProcessUniverse.java]: Comments.

File: it/src/main/java/org/corfudb/universe/universe/vm/VmConfigFileUtil.java [hunk 23]
```diff
@@ -0,0 +1,35 @@
+package org.corfudb.universe.universe.vm;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.Properties;
+
+public class VmConfigFileUtil {
```
Comment (vjeko) [inline it/src/main/java/org/corfudb/universe/universe/vm/VmConfigFileUtil.java]: Comments. Even by reading the class name, I have no idea what this class does. Maybe ConfigPropertiesLoader?

File: it/src/main/java/org/corfudb/universe/universe/vm/VmConfigFileUtil.java [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+    public static final String VM_PROPERTIES_FILE = "vm.properties";
+    public static final String VM_CREDENTIALS_PROPERTIES_FILE = "vm.credentials.properties";
+
+    private static Properties loadPropertiesFile(String propertiesFile) {
+        Properties credentials = new Properties();
+        URL credentialsUrl = ClassLoader.getSystemResource(propertiesFile);
+        try (InputStream is = credentialsUrl.openStream()) {
+            credentials.load(is);
+        } catch (IOException e) {
+            throw new IllegalStateException("Can't load credentials", e);
```
Comment (vjeko) [inline it/src/main/java/org/corfudb/universe/universe/vm/VmConfigFileUtil.java]: The exception is not correct.

File: it/src/main/resources/logback.xml [hunk 25]
```diff
@@ -24,6 +26,6 @@
 
     <!-- Control logging levels for individual components here. -->
     <root level="info">
-        <!--<appender-ref ref="STDOUT"/>-->
```
Comment (vjeko) [inline it/src/main/resources/logback.xml]: Do you just want to remove this?

File: it/src/test/java/org/corfudb/universe/GenericIntegrationTest.java [hunk 26]
```diff
@@ -1,111 +1,70 @@
...
 import org.junit.Before;
 import org.junit.Rule;
 import org.junit.rules.TestName;
 
-import java.util.Set;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.function.Consumer;
 
 public abstract class GenericIntegrationTest {
```
Comment (vjeko) [inline it/src/test/java/org/corfudb/universe/GenericIntegrationTest.java]: Comments.  What is this suppose to test?

File: it/src/test/java/org/corfudb/universe/scenario/ConcurrentClusterResizeIT.java [hunk 27]
```diff
@@ -34,9 +34,16 @@ public void concurrentClusterResizeTest() {
         // Deploy a five nodes cluster
         final int numNodes = 5;
 
-        getScenario(numNodes).describe((fixture, testCase) -> {
-            ClientParams clientFixture = fixture.getClient();
-            CorfuCluster corfuCluster = universe.getGroup(fixture.getCorfuCluster().getName());
+        workflow(wf -> {
+            wf.setupDocker(fixture -> fixture.getCluster().numNodes(numNodes));
+            wf.setupProcess(fixture -> fixture.getCluster().numNodes(numNodes));
+            wf.setupVm(fixture -> fixture.getCluster().numNodes(numNodes));
```
Comment (vjeko) [inline it/src/test/java/org/corfudb/universe/scenario/ConcurrentClusterResizeIT.java]: I don't get this. Why are we setting up all three?

File: it/src/test/java/org/corfudb/universe/GenericIntegrationTest.java [hunk 28]
```diff
@@ -1,111 +1,70 @@
...
+        public String getAppVersion() {
+            String version = System.getProperty("project.version");
+            if (version != null && !version.isEmpty()) {
+                return version;
+            }
+
+            return parseAppVersionInPom();
+        }
 
-    public Scenario<UniverseParams, AbstractUniverseFixture<UniverseParams>>getScenario(
```
Comment (vjeko) [inline it/src/test/java/org/corfudb/universe/GenericIntegrationTest.java]: How are the metrics port suppose to be passed? Right now this breaks all tests that rely on this feature. Here it's how it is being initialized right now:

```
        getScenario(3, Collections.singleton(metricsPort)).describe((fixture, testCase) -> {
            CorfuCluster corfuCluster = universe.getGroup(fixture.getCorfuCluster().getName());
            CorfuClient corfuClient = corfuCluster.getLocalCorfuClient(metricsPort);
```
What is the new workflow suppose to be?

File: benchmarks/src/main/java/org/corfudb/benchmarks/util/DataUnit.java (hunk 4 above)
Comment (xnull) [reply benchmarks/src/main/java/org/corfudb/benchmarks/util/DataUnit.java]: I don't see the reasons for that.
I have found https://github.com/sebhoss/storage-units on github. It has 4 stars. Also my implementation is just 38 LOC. Also, I'm not sure that the 3rd party library in maven central.  And I really don't the idea to a dependency to an unknown library

File: checks.xml (hunk 6 above)
Comment (xnull) [reply checks.xml]: All these checks generate a too huge number of warnings in the log.  
No one actually going to spend his time fixing that, but it makes the log analysis much harder.

I see that it can be useful only when we enable and make this check required.

Also, I have deleted the checkstyle javadoc checks. 
Overall it generated so many warning which prevented travis-ci to build the project.
It exceeds the 10k lines limit in Travis-ci.

File: checks.xml (hunk 7 above)
Comment (xnull) [reply checks.xml]: see above
The check is useless and generated too many warnings

File: checks.xml (hunk 8 above)
Comment (xnull) [reply checks.xml]: see above

File: it/src/main/java/org/corfudb/universe/node/server/CorfuServerParams.java (hunk 11 above)
Comment (xnull) [reply it/src/main/java/org/corfudb/universe/node/server/CorfuServerParams.java]: It's more flexible, it allows you to use different versions of corfu servers in one cluster. That could be the case very soon. That's why I decided to have this parameter in ClusterParams and in the ServerParams

File: it/src/main/java/org/corfudb/universe/node/server/CorfuServerParams.java (hunk 12 above)
Comment (xnull) [reply it/src/main/java/org/corfudb/universe/node/server/CorfuServerParams.java]: it is already configurable
```
wf.setupDocker(fixture -> {
    fixture.getServer(). serverJarDirectory("my_custom_jar_directory");
});
```

File: it/src/main/java/org/corfudb/universe/node/server/process/CorfuProcessManager.java (hunk 14 above)
Comment (xnull) [reply it/src/main/java/org/corfudb/universe/node/server/process/CorfuProcessManager.java]: done

File: .mvn/wrapper/maven-wrapper.properties (hunk 1 above)
Comment (xnull) [reply .mvn/wrapper/maven-wrapper.properties]: done

File: benchmarks/src/jmh/java/org/corfudb/benchmarks/cluster/ClusterBenchmark.java (hunk 2 above)
Comment (xnull) [reply benchmarks/src/jmh/java/org/corfudb/benchmarks/cluster/ClusterBenchmark.java]: done

File: benchmarks/src/jmh/resources/vm.properties (hunk 3 above)
Comment (xnull) [reply benchmarks/src/jmh/resources/vm.properties]: done

File: benchmarks/src/main/resources/logback.xml (hunk 5 above)
Comment (xnull) [reply benchmarks/src/main/resources/logback.xml]: done

File: it/src/main/java/org/corfudb/universe/UniverseManager.java (hunk 9 above)
Comment (xnull) [reply it/src/main/java/org/corfudb/universe/UniverseManager.java]: done

File: it/src/main/java/org/corfudb/universe/UniverseManager.java (hunk 10 above)
Comment (xnull) [reply it/src/main/java/org/corfudb/universe/UniverseManager.java]: done

File: it/src/main/java/org/corfudb/universe/node/server/process/ExecutionHelper.java (hunk 16 above)
Comment (xnull) [reply it/src/main/java/org/corfudb/universe/node/server/process/ExecutionHelper.java]: if it doesn't it will throw an exception

File: checks.xml (hunk 6 above)
Comment (vjeko) [reply checks.xml]: Please put this into a separate commit.

Comment (xnull) [reply checks.xml]: I can't, if I leave it as is I wouldn't be able to build the project.
When we deploy dependencies we also deploy source/javadoc from now. 
And checkstyle started to generate too many warnings, I had to disable those checks.

Comment (vjeko) [reply checks.xml]: You are claiming that you are unable to build the project -- warnings aside -- what is the problem? If you are running into Checkstyle compliance issues, then the right solution is to fix your code.

File: it/src/main/resources/logback.xml (hunk 25 above)
Comment (xnull) [reply it/src/main/resources/logback.xml]: fixed

File: it/src/main/java/org/corfudb/universe/universe/vm/VmConfigFileUtil.java (hunk 24 above)
Comment (xnull) [reply it/src/main/java/org/corfudb/universe/universe/vm/VmConfigFileUtil.java]: fixed

File: it/src/main/java/org/corfudb/universe/universe/vm/VmConfigFileUtil.java (hunk 23 above)
Comment (xnull) [reply it/src/main/java/org/corfudb/universe/universe/vm/VmConfigFileUtil.java]: done (doc + renamed)

File: it/src/main/java/org/corfudb/universe/universe/process/ProcessUniverse.java (hunk 22 above)
Comment (xnull) [reply it/src/main/java/org/corfudb/universe/universe/process/ProcessUniverse.java]: done

File: it/src/main/java/org/corfudb/universe/scenario/fixture/Fixtures.java (hunk 21 above)
Comment (xnull) [reply it/src/main/java/org/corfudb/universe/scenario/fixture/Fixtures.java]: done

File: it/src/main/java/org/corfudb/universe/scenario/fixture/FixtureUtil.java (hunk 20 above)
Comment (xnull) [reply it/src/main/java/org/corfudb/universe/scenario/fixture/FixtureUtil.java]: done

File: it/src/main/java/org/corfudb/universe/node/server/process/ProcessCorfuServer.java (hunk 19 above)
Comment (xnull) [reply it/src/main/java/org/corfudb/universe/node/server/process/ProcessCorfuServer.java]: fixed

File: it/src/main/java/org/corfudb/universe/node/server/process/ProcessCorfuServer.java (hunk 18 above)
Comment (xnull) [reply it/src/main/java/org/corfudb/universe/node/server/process/ProcessCorfuServer.java]: fixed

File: it/src/main/java/org/corfudb/universe/node/server/process/ProcessCorfuServer.java (hunk 17 above)
Comment (xnull) [reply it/src/main/java/org/corfudb/universe/node/server/process/ProcessCorfuServer.java]: It is an IP address, not a dns name.
the constant has extracted.

File: it/src/main/java/org/corfudb/universe/node/server/process/ExecutionHelper.java (hunk 15 above)
Comment (xnull) [reply it/src/main/java/org/corfudb/universe/node/server/process/ExecutionHelper.java]: done.
it's on a local machine (added that info into ExecutionHelper javadoc).

File: it/src/main/java/org/corfudb/universe/node/server/CorfuServerParams.java (hunk 12 above)
Comment (xnull) [reply it/src/main/java/org/corfudb/universe/node/server/CorfuServerParams.java]: Actually, everything that you can see in any `Prop` classes can be configured in your test

File: it/src/test/java/org/corfudb/universe/scenario/ConcurrentClusterResizeIT.java (hunk 27 above)
Comment (xnull) [reply it/src/test/java/org/corfudb/universe/scenario/ConcurrentClusterResizeIT.java]: 1. In `UniverseManager` you define what type of the universe you would like to have.
2. in the test you don't know what type of the universe is going to be used
3. we define how many nodes we want to have for all possible types of the universe. In general, there could be different settings for different types of universes. The fixtures of the different universes can have completely different fields and settings, so you have to define a parameter like `numNodes` for all universes in this case.

And yes, we can provide common settings across all universes, like `numNodes`, but I wouldn't like to do that now. For now, it's not clear what settings I'd like to make  common across different universes

File: it/src/test/java/org/corfudb/universe/GenericIntegrationTest.java (hunk 28 above)
Comment (xnull) [reply it/src/test/java/org/corfudb/universe/GenericIntegrationTest.java]:
```
wf.setupDocker(fixture -> {
                //fixture.getSupportServer(). define whatever you need
                fixture.getSupportServer().metricPorts(metricPorts);
            });
...
```

File: checks.xml (hunk 6 above)
Comment (xnull) [reply checks.xml]: It's not my code,
It's thousands of thousands of warnings (javadoc + imports) across all projects in the Corfu.

Comment (xnull) [reply checks.xml]: The reason why it wasn't broken before is that before we build only jar files, now we also build javadoc+sources for you modules

Comment (xnull) [reply checks.xml]: And when you build jar+javadoc+sources it generate 3x more warnings.
I believe having sources+javadoc for your artifact is way more important than having checkstyle warning that everyone ignores

Comment (vjeko) [reply checks.xml]: Can you please be more specific. What exactly is broken? Does the code not compile?

I checked-out your branch and did `git checkout HEAD~10 ./checks.xml`. I am not seeing any compilation issues.

File: it/src/test/java/org/corfudb/universe/GenericIntegrationTest.java (hunk 26 above)
Comment (xnull) [reply it/src/test/java/org/corfudb/universe/GenericIntegrationTest.java]: done

File: checks.xml (hunk 6 above)
Comment (xnull) [reply checks.xml]: The build on `travis-ci` is broken. 10k log limit.
The code is ok, you can compile it, it passes or the tests.

```
Also, I have deleted the checkstyle javadoc checks.
Overall it generated so many warning which prevented travis-ci to build the project.
It exceeds the 10k lines limit in Travis-ci.
```

Comment (vjeko) [reply checks.xml]: Looking at the recent Travis builds and test runs, I am not seeing any failures. Can you please point me to specific Travis CI failures that are caused by verbose output? For example, here is one of the recent PRs: https://travis-ci.org/CorfuDB/CorfuDB/builds/600908524

The question about check-style enforcement and log verbosity is something we should look into _outside_ of this patch. I see you are already passing `-q` flag to the Travis pipeline, which reduces log verbosity.

Comment (xnull) [reply checks.xml]: Again:
1. Corfu build in my branch generates much more logs, because I build javadoc+sources for the jar artifacts
2. You won't see the issue in other branches, see point 1.
3. this build https://travis-ci.org/CorfuDB/CorfuDB/jobs/597484304

what I did:
1. changed checkstyle
2. added `-q` parameter, to reduce the log https://github.com/CorfuDB/CorfuDB/blob/0ff85b92982ca2ec8f0fd214fbf4e1b3efb30aa7/.travis.yml#L19

Comment (xnull) [reply checks.xml]: Also if you disable all the logs what is the point for the warnings?
We just disable everything, it's even more than disable a couple of checks

P.S. yes I understand the diff between checkstyle and warnings in the log, but still

Comment (xnull) [reply checks.xml]: As discussed offline, we bring those checks back and double-check that it causes the problem. If not then we don't change check style. If yes, will figure out what to do

File: pom.xml.versionsBackup [hunk 29]
```diff
@@ -0,0 +1,464 @@
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
```
Comment (vjeko) [inline pom.xml.versionsBackup]: Delete this file.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java [hunk 30]
```diff
@@ -186,9 +183,9 @@ public int hashCode() {
      * @param <V>           Value type
      * @return              A type token to pass to the builder.
      */
-    static <K, V> TypeToken<CorfuTable<K, V>>
-        getMapType() {
-            return new TypeToken<CorfuTable<K, V>>() {};
+    public static <K, V> TypeToken<CorfuTable<K, V>> getMapType() {
+        return new TypeToken<CorfuTable<K, V>>() {
+        };
```
Comment (vjeko) [inline runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java]: Can you please revert this file. It will cause bunch of conflicts with my changes.

**Comment by vjeko** [comment]
Looks good. I tested the metrics part, and there are no regressions. Just make sure to un-stage CorfuTable and pom.xml.versionsBackup files.
